atspi-common/src/object_match.rs

```diff
@@ -0,0 +1,281 @@
+#[derive(Debug, Clone, Default)]
+pub struct ObjectMatchRuleBuilder {
+	states: StateSet,
+	states_mt: MatchType,
+	attr: HashMap<String, String>,
+	attr_mt: MatchType,
+	roles: Vec<Role>,
+	roles_mt: MatchType,
+	ifaces: InterfaceSet,
+	ifaces_mt: MatchType,
+	invert: bool,
+}
+
+impl ObjectMatchRuleBuilder {
+	/// Insert a `StateSet` to the builder
+	#[must_use]
+	pub fn states(mut self, state_set: StateSet, mt: MatchType) -> Self {
+		self.states = state_set;
+		self.states_mt = mt;
+		self
+	}
+
+	/// Insert a map of attributes
+	#[must_use]
+	pub fn attributes(mut self, attributes: HashMap<String, String>, mt: MatchType) -> Self {
+		self.attr = attributes;
+		self.attr_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Role`s
+	#[must_use]
+	pub fn roles(mut self, roles: &[Role], mt: MatchType) -> Self {
+		self.roles = roles.into();
+		self.roles_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Interface`s
+	#[must_use]
+	pub fn interfaces(mut self, interfaces: &[Interface], mt: MatchType) -> Self {
+		self.ifaces = interfaces.iter().copied().collect::<InterfaceSet>();
+		self.ifaces_mt = mt;
+		self
+	}
+
+	/// Sets the inversion of the `MatchRule`, defaults to `false`, no inversion.
+	#[must_use]
+	pub fn invert(mut self, invert: bool) -> Self {
+		self.invert = invert;
+		self
+	}
+
+	/// Builds the [`ObjectMatchRule`]
+	///
+	/// [`ObjectMatchRule``]: crate::object_match::ObjectMatchRule
+	#[must_use]
+	pub fn build(self) -> ObjectMatchRule {
+		ObjectMatchRule {
+			states: self.states,
+			states_mt: self.states_mt,
+			attr: self.attr,
+			attr_mt: self.attr_mt,
+			roles: self.roles,
+			roles_mt: self.roles_mt,
+			ifaces: self.ifaces,
+			ifaces_mt: self.ifaces_mt,
+			invert: self.invert,
+			phantom: PhantomData,
+		}
+	}
+}
+
+/// Enumeration used by [`ObjectMatchRule`] to specify how to interpret [`ObjectRef`] objects.
+///
+/// [`ObjectRef`]: crate::object_ref::ObjectRef
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Type, Default)]
+#[repr(i32)]
+pub enum MatchType {
+	#[default]
+	/// Invalidates match criterion.
+	Invalid,
+
+	/// All of the criteria must be met.
+	All,
+
+	/// Any of the criteria must criteria must be met.
+	Any,
+
+	/// None of the criteria must be met.
+	NA,
+
+	/// Same as [`Self::All`] if the criterion item is non-empty - All of the criteria must be met.
+	/// For empty criteria this rule requires the returned value to also have empty set.
+	Empty,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Type)]
+#[repr(u32)]
+/// Enumeration used by interface `CollectionProxy` to specify the way [`ObjectRef`]
+/// objects should be sorted.
+///
+/// # Note
+///
+/// Only `Canonical` and `ReverseCanonical` are currently implemented.
+///
+/// see: [issue 140. "Are all the AtspiCollectionSortOrder types really a thing?"](https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/140)
+///
+/// [`ObjectRef`]: crate::object_ref::ObjectRef
+pub enum SortOrder {
```

Comment:
Another question: what happens if we run code that uses one of these unimplemented sort orders? Does it fail with a DBus message or send zero items? If it fails, I feel like there must be a better way to restrict usage of the variants that we _already_ know will instantly fail: I'm thinking a different-but-similar solution to `#[deprecated]` that will warn automatically when you use a marker variant of an enum.